docs/protocols/xlang_object_graph_spec.md

```diff
@@ -0,0 +1,635 @@
+- Read multiple bytes as an int/long, then split into multiple bytes to reduce memory IO and bound check cost.
+- Try to use one varint/long to write flags and length together to save one byte cost and reduce memory io.
+- Condition branches are less expensive compared to memory IO cost unless there are too many branches.
+
+### Fast deserialization for static languages without runtime codegen support
+
+For type evolution, the serializer will encode the type meta into the serialized data. The deserializer will compare
+this meta with class meta in current process, and use the diff to determine how to deserialize the data.
+
+For java/javascript/python, we can use the diff to generate serializer code at runtime and load it as class/function for
+deserialization. In this way, the type evolution will be as fast as type consist mode.
+
+For C++/Rust, we can't generate the serializer code at runtime. So we need to generate the code at compile-time using
+meta programing. But at that time, we don't know the type schema in other processes, so we can't generate the serializer
+code for such inconsistent types. We may need to generate the code which has a loop and compare field name one by one to
+decide whether deserialize and assign the field or skip the field value.
+
+One fast way is that we can optimize the string comparison into `jump` instructions:
+
+- Assume current type has `n` fields, peer type has `n1` fields.
+- Generate an auto growing `field id` from `0` for every sorted field in current type at the compile time.
+- Compare the received type meta with current type, generate same id if the field name is same, otherwise generate an
+  auto growing id starting from `n`, cache this meta at runtime.
+- Iterate the fields of received type meta, use a `switch` to compare the `field id` to deserialize data
+  and `assign/skip` field value. **Continuous** field id will be optimized into `jump` in `switch` block, so it will
+  very fast.
+
+Here is an example, suppose process A has a class `Foo` with version 1 defined as `Foo1`, process B has a class `Foo`
+with version 2 defined as `Foo2`:
+
+```c++
+// class Foo with version 1
+class Foo1 {
+  int32_t v1; // id 0
+  std::string v2; // id 1
+}
+// class Foo with version 2
+class Foo2 {
+  // id 0, but will have id 2 in process A
+  bool v0;
+  // id 1, but will have id 0 in process A
+  int32_t v1;
+  // id 2, but will have id 3 in process A
+  int64_t long_value;
+  // id 3, but will have id 1 in process A
+  std::string v2;
+  // id 4, but will have id 4 in process A
+  std::vector<std::string> list;
+}
+```
+
+When process A received serialized `Foo2` from process B, here is how it deserialize the data:
+
+```c++
+Foo1 &foo1 = xxx;
+std::vector<fury::FieldInfo> &field_infos = type_meta.field_infos;
+for (auto &field_info : field_infos) {
+  switch (field_info.field_id) {
+    case 0:
+      foo1.v1 = buffer.read_varint32();
```

Comment:
In this sample,  the field v1 in Foo1 assigned ` id 0` and it's type was int32, whereas in Foo2 , `id 0 ` refers to v0 which is a boolean. how did `case0:   foo1.v1 = buffer.read_varint32();` worked correctly  under these conditions?

Reply:
The id won't be written in the serialized data, the whole type meta will carry the field name, so when deserialization, fury can compare the field name and generate a different id for fields not match

Reply:
Ok, i got it